File: src/gitscale/commands/fetch.py
"""Fetch latest remote state without modifying local files."""

from pathlib import Path

import click

from gitscale.config import ConfigError, RepoEntry, find_config, load_config
from gitscale.git import GitError, fetch_repo
from gitscale.storage import StorageError, fetch_manifest


@click.command()
@click.option(
    "-C",
    "--root",
    type=click.Path(exists=True, path_type=Path),
    default=None,
    help="Root directory containing .gitscale.toml (default: auto-detect).",
)
@click.argument("names", nargs=-1)
@click.pass_context
def fetch(
    ctx: click.Context,
    root: Path | None,
    names: tuple[str, ...],
) -> None:
    """Fetch latest remote state for sub-repositories.

    For git repos: runs git fetch.
    For manifests: checks remote ETag via HEAD request.
    If NAMES are given, fetch only those entries. Otherwise fetch all.
    """
    verbose: bool = ctx.obj["verbose"]

    try:
        config_path = find_config(root)
    except ConfigError as e:
        raise click.ClickException(str(e)) from None

    config_root = config_path.parent
    config = load_config(config_path)
    selected = _filter_entries(config.repos, names)

    if not selected:
        click.echo("Nothing to fetch.")
        return

    failed = 0
    for entry in selected:
        try:
            if entry.is_manifest:
                if not config.storage_url:
                    click.echo(
                        f"  FAIL  {entry.directory}: no [storage] configured",
                        err=True,
                    )
                    failed += 1
                    continue
                dest = config_root / entry.directory
                revision = entry.revision or "HEAD"
                result = fetch_manifest(
                    config.storage_url, entry.repo_url, revision, dest
                )
                if result.exists:
                    click.echo(f"  ok    {entry.directory} (manifest)")
                else:
                    click.echo(
                        f"  skip  {entry.directory} (no remote manifest)"
                    )
                continue

            dest = config_root / entry.directory
            if not dest.exists():
                click.echo(f"  skip  {entry.directory} (not cloned)")
                continue
            if verbose:
                click.echo(f"  fetch {entry.directory}")
            fetch_repo(entry, config_root)
            click.echo(f"  ok    {entry.directory}")
        except (GitError, StorageError) as e:
            click.echo(f"  FAIL  {entry.directory}: {e}", err=True)
            failed += 1

    if failed:
        raise click.ClickException(f"{failed} repo(s) failed to fetch")


def _filter_entries(
    entries: list[RepoEntry], names: tuple[str, ...]
) -> list[RepoEntry]:
    if not names:
        return entries
    name_set = set(names)
    matched = [e for e in entries if e.directory in name_set]
    unknown = name_set - {e.directory for e in matched}
    if unknown:
        raise click.ClickException(
            f"Unknown repos: {', '.join(sorted(unknown))}"
        )
    return matched

File: src/commands/fetch.rs
use anyhow::Result;
use std::collections::HashMap;
use std::io::Write;
use std::path::Path;

use crate::cache;
use crate::commands::cache as cache_cmd;
use crate::commands::clone::filter_entries;
use crate::config::{find_config, load_config};
use crate::git::{fetch_repo, is_ci};
use crate::progress::{run_parallel, RepoStatus};
use crate::share;
use crate::storage::fetch_artefact;

pub fn run(
    root: Option<&Path>,
    names: &[String],
    verbose: bool,
    no_cache: bool,
    interactive: bool,
    out: &mut dyn Write,
    err: &mut dyn Write,
) -> Result<()> {
    let config_path = find_config(root)?;
    let config_root = config_path.parent().unwrap().to_path_buf();
    let config = load_config(&config_path)?;
    let selected = filter_entries(&config.repos, names)?;

    if selected.is_empty() {
        writeln!(out, "Nothing to fetch.")?;
        return Ok(());
    }

    let entry_map: HashMap<&str, &crate::config::RepoEntry> = selected
        .iter()
        .map(|e| (e.directory.as_str(), e))
        .collect();
    let dir_names: Vec<String> = selected.iter().map(|e| e.directory.clone()).collect();
    let storage_url = &config.storage_url;
    let ci = is_ci();
    let workspace = share::source_workspace(&config_root);
    let dissociate = config.share.dissociate;
    let cache = cache_cmd::open(&config, no_cache);

    let failed = run_parallel(
        "Fetching...",
        &dir_names,
        interactive,
        |name| {
            let entry = &entry_map[name];

            if entry.is_artefact() {
                if storage_url.is_empty() {
                    return RepoStatus::Fail(format!("{}: no [storage] configured", name));
                }
                let dest = config_root.join(&entry.directory);
                let revision = if entry.revision.is_empty() {
                    "HEAD"
                } else {
                    &entry.revision
                };
                return match fetch_artefact(storage_url, &entry.repo_url, revision, &dest) {
                    Ok(result) => {
                        if result.exists {
                            RepoStatus::Ok(format!("{} (artefact)", name))
                        } else {
                            RepoStatus::Skip(format!("{} (no remote artefact)", name))
                        }
                    }
                    Err(e) => RepoStatus::Fail(format!("{}: {}", name, e)),
                };
            }

            let dest = config_root.join(&entry.directory);
            if !dest.exists() {
                return RepoStatus::Skip(format!("{} (not cloned)", name));
            }

            // Updates the cache entry as well: a fetch that only advanced
            // this workspace's refs would leave every other one to download
            // the same objects again.
            let from = cache::source_for(
                cache.as_ref(),
                workspace.as_deref(),
                entry,
                dissociate,
                ci,
                verbose,
            );
            match fetch_repo(entry, &config_root, &from) {
                Ok(()) => RepoStatus::Ok(name.to_string()),
                Err(e) => RepoStatus::Fail(format!("{}: {}", name, e)),
            }
        },
        out,
        err,
    )?;

    if failed > 0 {
        anyhow::bail!("{} repo(s) failed to fetch", failed);
    }
    Ok(())
}
